File: rd-kt/rd-framework/src/commonMain/kotlin/com/jetbrains/rd/framework/impl/SwitchingViewableSet.kt
package com.jetbrains.rd.framework.impl

import com.jetbrains.rd.util.addUnique
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.lifetime.SequentialLifetimes
import com.jetbrains.rd.util.reactive.AddRemove
import com.jetbrains.rd.util.reactive.IMutableViewableSet
import com.jetbrains.rd.util.reactive.ISource
import com.jetbrains.rd.util.reactive.IViewableSet

class SwitchingViewableSet<T : Any>(lifetime: Lifetime, private var myBackingSet: IMutableViewableSet<T>): IMutableViewableSet<T> {

    private val myListeners = LinkedHashSet<(IViewableSet.Event<T>) -> Unit>()
    private val myAdviseSeqLifetimes = SequentialLifetimes(lifetime)

    override val change: ISource<IViewableSet.Event<T>>
        get() = this

    init {
        adviseForBackingSet()
    }

    private fun adviseForBackingSet() {
        val adviseLt = myAdviseSeqLifetimes.next()
        var initial = true
        myBackingSet.advise(adviseLt) { event ->
            if (!initial) myListeners.forEach { it(event) }
        }
        initial = false
    }

    override fun advise(lifetime: Lifetime, handler: (IViewableSet.Event<T>) -> Unit) {
        myBackingSet.forEach { handler(IViewableSet.Event(AddRemove.Add, it)) }
        myListeners.addUnique(lifetime, handler)
    }

    fun changeBackingSet(newBackingSet: IMutableViewableSet<T>, isNewSetMaster: Boolean = true) {
        if (isNewSetMaster) {
            val missingValues = myBackingSet - newBackingSet
            val newValue = newBackingSet - myBackingSet
            missingValues.forEach {
                myListeners.forEach { listener ->
                    listener(IViewableSet.Event(AddRemove.Remove, it))
                }
            }
            newValue.forEach {
                myListeners.forEach { listener ->
                    listener(IViewableSet.Event(AddRemove.Add, it))
                }
            }
        }
        val oldBackingSet = myBackingSet
        myBackingSet = newBackingSet
        if(!isNewSetMaster) {
            newBackingSet.retainAll(oldBackingSet)
            newBackingSet.addAll(oldBackingSet)
        }
        adviseForBackingSet()
    }


    override fun add(element: T) = myBackingSet.add(element)
    override fun addAll(elements: Collection<T>) = myBackingSet.addAll(elements)
    override fun clear() = myBackingSet.clear()
    override fun iterator() = myBackingSet.iterator()
    override fun remove(element: T) = myBackingSet.remove(element)
    override fun removeAll(elements: Collection<T>) = myBackingSet.removeAll(elements)
    override fun retainAll(elements: Collection<T>) = myBackingSet.removeAll(elements)
    override val size: Int
        get() = myBackingSet.size
    override fun contains(element: T) = myBackingSet.contains(element)
    override fun containsAll(elements: Collection<T>) = myBackingSet.containsAll(elements)
    override fun isEmpty() = myBackingSet.isEmpty()
}

File: rd-net/RdFramework/Impl/SwitchingViewableSet.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using JetBrains.Collections.Viewable;
using JetBrains.Lifetimes;

namespace JetBrains.Rd.Impl
{
  public class SwitchingViewableSet<T> : IViewableSet<T>
  {
    private IViewableSet<T> myBackingSet;
    private LifetimedList<Action<SetEvent<T>>> myListeners = new LifetimedList<Action<SetEvent<T>>>();
    private readonly SequentialLifetimes myAdviseLifetimes;

    public SwitchingViewableSet(Lifetime lifetime, IViewableSet<T> backingSet)
    {
      myBackingSet = backingSet;
      myAdviseLifetimes = new SequentialLifetimes(lifetime);
      
      AdviseForBackingSet();
    }

    public void ChangeBackingSet(IViewableSet<T> newBackingSet, bool isNewSetMaster = true)
    {
      if (isNewSetMaster)
      {
        var missingValues = myBackingSet.Except(newBackingSet);
        var newValues = newBackingSet.Except(myBackingSet);
        
        foreach (var missingValue in missingValues) 
          FireListeners(SetEvent<T>.Remove(missingValue));
        foreach (var newValue in newValues) 
          FireListeners(SetEvent<T>.Add(newValue));
      }

      var oldBackingSet = myBackingSet;
      myBackingSet = newBackingSet;
      if (!isNewSetMaster)
      {
        var valuesToRemove = newBackingSet.Where(it => !oldBackingSet.Contains(it)).ToList();
        foreach (var newValue in valuesToRemove)
          newBackingSet.Remove(newValue);

        foreach (var oldValue in oldBackingSet) 
          newBackingSet.Add(oldValue);
      }
      AdviseForBackingSet();
    }

    private void AdviseForBackingSet()
    {
      var lt = myAdviseLifetimes.Next();
      var initial = true;
      myBackingSet.Advise(lt, setEvent =>
      {
        // ReSharper disable once AccessToModifiedClosure
        if (initial) return;
        FireListeners(setEvent);
      });
      initial = false;
    }

    private void FireListeners(SetEvent<T> setEvent)
    {
      foreach (var (lifetime, value) in myListeners)
      {
        if (!lifetime.IsAlive) continue;
        value?.Invoke(setEvent);
      }
    }

    public void Advise(Lifetime lifetime, Action<SetEvent<T>> handler)
    {
      foreach (var existingValue in myBackingSet) 
        handler(SetEvent<T>.Add(existingValue));

      myListeners.Add(lifetime, handler);
    }

    public ISource<SetEvent<T>> Change => this;

    #region IViewableSet delegation
    
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    public IEnumerator<T> GetEnumerator() => myBackingSet.GetEnumerator();
    public void Add(T item) => myBackingSet.Add(item);
    public void Clear() => myBackingSet.Clear();
    public bool Contains(T item) => myBackingSet.Contains(item);
    public void CopyTo(T[] array, int arrayIndex) => myBackingSet.CopyTo(array, arrayIndex);
    public bool Remove(T item) => myBackingSet.Remove(item);
    public int Count => myBackingSet.Count;
    public bool IsReadOnly => myBackingSet.IsReadOnly;
    #if !NET35
    bool ISet<T>.Add(T item) => myBackingSet.Add(item);
    public void ExceptWith(IEnumerable<T> other) => myBackingSet.ExceptWith(other);
    public void IntersectWith(IEnumerable<T> other) => myBackingSet.IntersectWith(other);
    public bool IsProperSubsetOf(IEnumerable<T> other) => myBackingSet.IsProperSubsetOf(other);
    public bool IsProperSupersetOf(IEnumerable<T> other) => myBackingSet.IsProperSupersetOf(other);
    public bool IsSubsetOf(IEnumerable<T> other) => myBackingSet.IsSubsetOf(other);
    public bool IsSupersetOf(IEnumerable<T> other) => myBackingSet.IsSupersetOf(other);
    public bool Overlaps(IEnumerable<T> other) => myBackingSet.Overlaps(other);
    public bool SetEquals(IEnumerable<T> other) => myBackingSet.SetEquals(other);
    public void SymmetricExceptWith(IEnumerable<T> other) => myBackingSet.SymmetricExceptWith(other);
    public void UnionWith(IEnumerable<T> other) => myBackingSet.UnionWith(other);
    #endif
    #endregion
  }
}